src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs

```diff
@@ -0,0 +1,226 @@
+                var message = "This project is not applicable for updating to CoreWCF since references to WCF were not found. No more work needs to be done and this step is complete.";
+                Logger.LogInformation(message);
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, message, BuildBreakRisk.Low));
+            }
+
+            // construct updaters
+            _configUpdater = UpdaterFactory.GetConfigUpdater(_path["config"].First(), Logger);
+            _packageUpdater = UpdaterFactory.GetPackageUpdater(_path["proj"].First(), Logger);
+            if (_path.ContainsKey("directives"))
+            {
+                _directiveUpdaters = UpdaterFactory.GetDirectiveUpdaters(_path["directives"], Logger);
+            }
+
+            if (_configUpdater is not null)
+            {
+                var configContext = UpdateRunner.GetContext(_configUpdater);
+                _sourceCodeUpdater = UpdaterFactory.GetSourceCodeUpdater(_path["main"].First(), configContext, Logger);
+            }
+
+            if (_configUpdater is null || _packageUpdater is null || _sourceCodeUpdater is null || (_path.ContainsKey("directives") && _directiveUpdaters is null))
+            {
+                Logger.LogWarning("Unexpected error happened when trying to construct the updaters. Please review error message and log.");
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Failed, "Updaters cannot be constructed. Please review error message and log for more information.", BuildBreakRisk.Medium));
+            }
+
+            return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, "WCF updaters are initialized. Ready to update.", BuildBreakRisk.None));
+        }
+
+        public Task<UpgradeStepApplyResult> Apply()
+        {
+            // Run updates
+            var projFile = UpdateRunner.PackageUpdate(_packageUpdater!, Logger);
+            var config = UpdateRunner.ConfigUpdate(_configUpdater!, Logger);
+            var source = UpdateRunner.SourceCodeUpdate(_sourceCodeUpdater!, Logger);
+            var directives = new List<SyntaxNode>();
+            if (_directiveUpdaters is not null)
+            {
+                directives = UpdateRunner.DirectiveUpdate(_directiveUpdaters, Logger);
+            }
+
+            // check for null changes
+            if (projFile is null || config is null || source is null || directives is null)
+            {
+                Logger.LogWarning("Unexpected error happened when trying to making changes to original files. Please review error message and log.");
+                return Task.FromResult(new UpgradeStepApplyResult(UpgradeStepStatus.Failed, "Project failed to port to CoreWCF. Please review error message and log for more information."));
+            }
+
+            // Write updates
+            UpdateRunner.WritePackageUpdate(projFile, _path!["proj"].First(), Logger);
```

Comment:
Rather than using `_path!` it would be better to check that `_path` is not null. It shouldn't ever happen, but who knows if a bug might creep in allowing Apply to be erroneously called without Initialize. If _path were null, you could just return a failing result. And if it's not, then you won't need !.